src/nvim/statusline.c

```diff
@@ -1120,15 +1124,28 @@ int build_stl_str_hl(win_T *wp, char *out, size_t outlen, char *fmt, int use_san
       itemisflag = true;
 
       if (reevaluate) {
-        fmt_p++;
+        if (numbercolumn && (fmt_p[1] == 'l' || fmt_p[1] == 'r')) {
+          reevaluate = false;
+        } else {
+          fmt_p++;
+        }
       }
 
       // Attempt to copy the expression to evaluate into
       // the output buffer as a null-terminated string.
       char *t = out_p;
       while ((*fmt_p != '}' || (reevaluate && fmt_p[-1] != '%'))
              && *fmt_p != NUL && out_p < out_end_p) {
-        *out_p++ = *fmt_p++;
+        if (numbercolumn && *fmt_p == '%' && (fmt_p[1] == 'l' || fmt_p[1] == 'r')) {
+          // Substitute '%l' and '%r' for 'numbercolumn'
+          int i = snprintf(out_p, (outlen - (size_t)(out_p - t)), "%ld",
+                           (long)get_vim_var_nr((fmt_p[1] == 'l')
+                                                ? VV_REDRAW_LNUM : VV_REDRAW_RELNUM));
+          out_p += i;
+          fmt_p += 2;
```

Comment:
> Hmm, just realized %l can't be used within %{ } so a v: variable will be necessary.

Still looking for feedback on whether we want to overload, and allow the substitution of `%r` and `%l` inside `STL_VIM_EXPR`. If so I could go ahead and update the docs to reflect that.
And if we allow substituting them, do we still want to set the `v:` varables? At that point we could store/access them in `win_T` too.

Reply:
Updated the docs under the assumption we want this, but still looking for feedback on whether we actually do.

Reply:
> And if we allow substituting them, do we still want to set the `v:` varables?

Let's add the `v:` vars no matter what, because they are generally useful. While overloading may be subject to later changes/debate.

Reply:
@justinmk by generally useful do you mean to say they should be set even when `'numbercolumn'`, or even `'number'` and `'relativenumber'` are unset? Or did you just mean generally useful in the context of `'numbercolumn'`?

Reply:
> by generally useful do you mean to say they should be set even when `'numbercolumn'`, or even `'number'` and `'relativenumber'` are unset? Or did you just mean generally useful in the context of `'numbercolumn'`?

The latter, i.e. the `v:` vars are not at risk of debate about overloading.